File: springboot_be/src/main/java/com/mady/springboot_be/services/UserService.java
package com.mady.springboot_be.services;

import java.nio.file.AccessDeniedException;
import java.util.List;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.ResponseEntity;

import com.mady.springboot_be.dettails.ApiResponse;
import com.mady.springboot_be.dtos.UserDTO;
import com.mady.springboot_be.exceptions.AdminDeletionException;
import com.mady.springboot_be.exceptions.ResourceNotFoundException;

/**
 * Service interface for user management operations.
 * 
 * Defines methods for:
 * - User CRUD operations (create, read, update, delete)
 * - Traditional and OAuth2 user creation
 * - Soft delete with recursive propagation to user data (advertisements,
 * products, images)
 * - ADMIN user protection (cannot delete the last ADMIN)
 * - Reactivation of soft-deleted users
 * - Pagination support for active/deleted users
 * 
 * @author Popa Madalina Mariana
 * @since Spring Boot 3.4.4 / Java 21
 */
public interface UserService {

    /**
     * Retrieves a user by email.
     * 
     * @param email the user's email
     * @return the UserDTO
     */
    UserDTO getUserByEmail(String email);

    /**
     * Counts active users.
     * 
     * @return count of active users
     * @throws ResourceNotFoundException if no users found
     */
    int countUsers() throws ResourceNotFoundException;

    /**
     * Retrieves paginated list of active users.
     * 
     * @param pageable pagination parameters
     * @return page of UserDTOs
     * @throws ResourceNotFoundException if no users found
     */
    Page<UserDTO> getUsers(Pageable pageable) throws ResourceNotFoundException;

    /**
     * Retrieves all active users as a list.
     * 
     * @return list of active UserDTOs
     */
    List<UserDTO> getAllActiveUsers();

    /**
     * Counts deleted (inactive) users.
     * 
     * @return count of deleted users
     * @throws ResourceNotFoundException if no deleted users found
     */
    int countDeletedUsers() throws ResourceNotFoundException;

    /**
     * Retrieves paginated list of deleted users.
     * 
     * @param pageable pagination parameters
     * @return page of UserDTOs
     * @throws ResourceNotFoundException if no deleted users found
     */
    Page<UserDTO> getDeletedUsers(Pageable pageable) throws ResourceNotFoundException;

    /**
     * Creates a new user (traditional registration with password).
     * 
     * @param userDTO the user data
     * @return the created UserDTO
     */
    UserDTO createUser(UserDTO userDTO);

    /**
     * Creates a new user from OAuth2 (Google) login (no password).
     * 
     * @param userDTO the user data
     * @return the created UserDTO
     */
    UserDTO createOAuthUser(UserDTO userDTO);

    /**
     * Retrieves a user by ID.
     * 
     * @param id the user ID
     * @return the UserDTO
     * @throws ResourceNotFoundException if user not found
     */
    UserDTO getUserById(Long id) throws ResourceNotFoundException;

    /**
     * Updates a user.
     * 
     * Prevents deactivation of the last ADMIN user.
     * 
     * @param id      the user ID
     * @param userDTO the updated user data
     * @return the updated UserDTO
     * @throws ResourceNotFoundException if user not found
     */
    UserDTO updateUser(Long id, UserDTO userDTO) throws ResourceNotFoundException;

    /**
     * Soft deletes a user (only non-ADMIN users).
     * 
     * @param userId the user ID
     * @return ResponseEntity with success/failure message
     * @throws ResourceNotFoundException if user not found
     * @throws AccessDeniedException     if trying to delete an ADMIN user
     */
    ResponseEntity<Object> deleteUser(Long userId)
            throws ResourceNotFoundException, IllegalArgumentException, AccessDeniedException;

    /**
     * Soft deletes all non-ADMIN users.
     * 
     * @return ResponseEntity with success/failure message
     * @throws AdminDeletionException if ADMIN users cannot be deleted
     */
    ResponseEntity<ApiResponse> deleteUsers() throws IllegalArgumentException, AdminDeletionException;

    /**
     * Reactivates all soft-deleted non-ADMIN users and their associated data.
     * 
     * @return ResponseEntity with success/failure message
     */
    ResponseEntity<ApiResponse> reactivateAllUsers();
}

File: angular_fe/src/app/services/path/user/user.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { UserDTO } from 'src/app/interfaces/dtos/user_dto.interface';
import { PageResponse } from 'src/app/interfaces/page/page_response.interface';
import { Pageable } from 'src/app/interfaces/page/pageable.interface';
import { ApiConfigService } from '../../api-config/api-config.service';

/**
 * @category Services
 *
 * @description
 * Handles CRUD operations for users, including retrieval of active and deleted users,
 * paginated queries, and bulk actions like reactivation or deletion.
 *
 * @example
 * ```typescript
 * constructor(private userService: UserService) {}
 *
 * // Get user by email
 * this.userService.getUserByEmail('test@example.com').subscribe(user => console.log(user));
 *
 * // Get paginated active users
 * this.userService.getUsers({ page: 0, size: 10 }).subscribe(page => console.log(page));
 *
 * // Delete a user
 * this.userService.deleteUser(123).subscribe(() => console.log('User deleted'));
 * ```
 *
 * @author Popa Madalina Mariana
 * @version 0.0.0
 */
@Injectable({
  providedIn: 'root',
})
export class UserService {
  /**
   * @description Initializes UserService.
   *
   * @param http Angular HttpClient for performing HTTP requests
   * @param apiConfig Service providing API endpoints
   */
  constructor(
    private http: HttpClient,
    private apiConfig: ApiConfigService,
  ) {}

  /**
   * @description Retrieves a user by email.
   *
   * @param email Email of the user
   * @returns Observable emitting the UserDTO
   */
  getUserByEmail(email: string): Observable<UserDTO> {
    return this.http.get<UserDTO>(`${this.apiConfig.apiUser}/email`, {
      params: new HttpParams().set('email', email),
    });
  }

  /**
   * @description Returns the total number of active users.
   *
   * @returns Observable emitting the count of active users
   */
  countUsers(): Observable<number> {
    return this.http.get<number>(`${this.apiConfig.apiUser}/count`);
  }

  /**
   * @description Retrieves a paginated list of active users.
   *
   * @param pageable Pageable object containing page, size, and optional sort
   * @returns Observable emitting a PageResponse of UserDTO
   */
  getUsers(pageable: Pageable): Observable<PageResponse<UserDTO>> {
    const params = new HttpParams()
      .set('page', pageable.page)
      .set('size', pageable.size)
      .set('sort', pageable.sort || '');

    return this.http.get<PageResponse<UserDTO>>(
      `${this.apiConfig.apiUser}/all`,
      { params },
    );
  }

  /**
   * @description Returns the total number of deleted users.
   *
   * @returns Observable emitting the count of deleted users
   */
  countDeletedUsers(): Observable<number> {
    return this.http.get<number>(`${this.apiConfig.apiUser}/count-deleted`);
  }

  /**
   * @description Retrieves a paginated list of deleted users.
   *
   * @param pageable Pageable object containing page, size, and optional sort
   * @returns Observable emitting a PageResponse of UserDTO
   */
  getDeletedUsers(pageable: Pageable): Observable<PageResponse<UserDTO>> {
    const params = new HttpParams()
      .set('page', pageable.page)
      .set('size', pageable.size)
      .set('sort', pageable.sort || '');

    return this.http.get<PageResponse<UserDTO>>(
      `${this.apiConfig.apiUser}/deleted`,
      { params },
    );
  }

  /**
   * @description Retrieves a user by their ID.
   *
   * @param userId ID of the user
   * @returns Observable emitting the UserDTO
   */
  getUserById(userId: number): Observable<UserDTO> {
    return this.http.get<UserDTO>(`${this.apiConfig.apiUser}/get/${userId}`);
  }

  /**
   * @description Updates an existing user.
   *
   * @param userId ID of the user to update
   * @param user Updated user data
   * @returns Observable emitting the updated UserDTO
   */
  updateUser(userId: number, user: UserDTO): Observable<UserDTO> {
    return this.http.put<UserDTO>(
      `${this.apiConfig.apiUser}/update/${userId}`,
      user,
    );
  }

  /**
   * @description Deletes a user by ID.
   *
   * @param userId ID of the user to delete
   * @returns Observable emitting the result
   */
  deleteUser(userId: number): Observable<any> {
    return this.http.delete(`${this.apiConfig.apiUser}/delete/${userId}`);
  }

  /**
   * @description Deletes all users except ADMIN.
   *
   * @returns Observable emitting void
   */
  deleteAllUsers(): Observable<void> {
    return this.http.delete<void>(`${this.apiConfig.apiUser}/delete/all`);
  }

  /**
   * @description Reactivates all users (bulk action).
   *
   * @returns Observable emitting the result
   */
  reactivateAllUsers(): Observable<any> {
    return this.http.put(`${this.apiConfig.apiUser}/reactivate/all`, {});
  }

  /**
   * @description Retrieves all active users (non-paginated).
   *
   * @returns Observable emitting an array of UserDTO
   */
  getAllActiveUsers(): Observable<UserDTO[]> {
    return this.http.get<UserDTO[]>(`${this.apiConfig.apiUser}/list/active`);
  }
}
